### PR TITLE
Oceanwater 609 autonomy listen to fault msgs

### PR DESCRIPTION
**Pull request by lanseafood**
Jira [609](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-609)
summary:
* all faults messages are being listened to by Autonomy
* reports on status of these messages during any autonomy call
* report is sent only once, and then resolved message is sent when fault is removed/unselected.

Test:
* run any roslaunch file
* run any autonomy plan
* go to RQT and select any fault
* observe the system level and component level report of fault (occurs once) in RED
* unselect fault
* observe the resolve message (occurs once)  for system and component level in WHITE

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,8 @@
 cmake_minimum_required(VERSION 2.8.3)
 project(ow_autonomy)
 
+set(EXEC_NAME autonomy)
```
Comment (kmdalal) [inline CMakeLists.txt:4]: Why was this needed?

File: CMakeLists.txt [hunk 2]
```diff
@@ -47,6 +51,7 @@ catkin_add_env_hooks(60.ow_autonomy
 
 
 include_directories(
+  include
```
Comment (kmdalal) [inline CMakeLists.txt:54]: Why was this line needed?

File: src/plexil-adapter/OwInterface.h [hunk 3]
```diff
@@ -97,6 +103,44 @@ class OwInterface
...
+  void antennaFaultCallback (const ow_faults::PTFaults::ConstPtr&);
+  bool checkFaultMessages(std::string fault_component, 
+                                        uint32_t msg_val, 
+                                        std::string key, 
+                                        uint32_t value, 
+                                        bool b );
+
+//////////////////// FAULTS FOR SYSTEM LEVEL ////////////////////////
+
+  std::map<std::string,std::pair<uint64_t, bool>> systemErrors =
```
Comment (kmdalal) [inline src/plexil-adapter/OwInterface.h:118]: I would have made these maps static variables in the .cpp file.  Was there a certain reason to make them class members, and (@Samahu ) is initializing maps in the header file idiomatic?  I always initialize member maps and other structures in the .cpp file.

**Comment by kmdalal** [comment]
I've added @Samahu as a reviewer.  I cannot start any nodes in the autonomy package:
```
ERROR: cannot launch node of type [ow_autonomy/autonomy_node]: Cannot locate node of type [autonomy_node] in package [ow_autonomy]. Make sure file exists in package path and permission is set to executable (chmod +x)
ERROR: cannot launch node of type [ow_autonomy/GuardedMoveServer]: Cannot locate node of type [GuardedMoveServer] in package [ow_autonomy]. Make sure file exists in package path and permission is set to executable (chmod +x)
```
and am wondering if it has to do with the cmake file edits, not all of which I understood.

**Comment by Samahu** [comment]
I was able to execute the described test:
```
[ INFO] [1614039096.825304992, 1916525002.707499981]:   Starting ROS service Grind
[ERROR] [1614039099.137517088, 1916525004.967499971]: SYSTEM ERROR: PT_EXECUTION_ERROR
[ERROR] [1614039099.137599726, 1916525004.967499971]: ANTENNA ERROR: HARDWARE_ERROR
[ INFO] [1614039108.585953543, 1916525010.967499971]: RESOLVED SYSTEM ERROR: PT_EXECUTION_ERROR
[ INFO] [1614039108.586018151, 1916525010.967499971]: RESOLVED ANTENNA ERROR: HARDWARE_ERROR
```

File: src/plexil-adapter/OwInterface.h (hunk 3 above)
Comment (lanseafood) [reply src/plexil-adapter/OwInterface.h:118]: We could make them static. I chose to make them a part of the class such that if ever we needed to recall this anywhere else in autonomy in the future.

File: CMakeLists.txt (hunk 2 above)
Comment (lanseafood) [reply CMakeLists.txt:54]: can remove

File: CMakeLists.txt (hunk 1 above)
Comment (lanseafood) [reply CMakeLists.txt:4]: will remove

File: src/plexil-adapter/OwInterface.h (hunk 3 above)
Comment (lanseafood) [reply src/plexil-adapter/OwInterface.h:118]: I placed the inifliazation in the .h folder to make the cpp cleaner since the values associated with each failure type would remain for the most part constant and the booleans are the only thing being updated if faults occur.

File: src/plexil-adapter/OwInterface.cpp [hunk 4]
```diff
@@ -310,6 +308,97 @@ static void handle_joint_fault (Joint joint, int joint_index,
...
+    uint64_t value = entry.second.first;
+    bool b = entry.second.second;
+
+    if (!b && ((msg_val & value) == value)){
+      ROS_ERROR("SYSTEM ERROR: %s", key.c_str() );
+      systemErrors[key].second = true;
+    }
+    else if (b && ((msg_val & value) != value)){
+      ROS_INFO("RESOLVED SYSTEM ERROR: %s", key.c_str() );
+      systemErrors[key].second = false;
```
Comment (Samahu) [inline src/plexil-adapter/OwInterface.cpp:329]: It does seem that you have a function `checkFaultMessages` that does the exact behavior which you applied in other parts but not .. is there a reason why you couldn't use that same function here?

File: src/plexil-adapter/OwInterface.h [hunk 5]
```diff
@@ -115,6 +159,11 @@ class OwInterface
   ros::Subscriber* m_rulSubscriber;
   ros::Subscriber* m_guardedMoveSubscriber;
 
+  ros::Subscriber* m_systemFaultMessagesSubscriber;
+  ros::Subscriber* m_armFaultMessagesSubscriber;
+  ros::Subscriber* m_powerFaultMessagesSubscriber;
+  ros::Subscriber* m_ptFaultMessagesSubscriber;
```
Comment (Samahu) [inline src/plexil-adapter/OwInterface.h:165]: Use [`std::unique_ptr`](https://en.cppreference.com/w/cpp/memory/unique_ptr) instead of raw pointers
```suggestion
  std::unique_ptr<ros::Subscriber> m_systemFaultMessagesSubscriber;
  std::unique_ptr<ros::Subscriber> m_armFaultMessagesSubscriber;
  std::unique_ptr<ros::Subscriber> m_powerFaultMessagesSubscriber;
  std::unique_ptr<ros::Subscriber> m_ptFaultMessagesSubscriber;
```

**Review by Samahu — CHANGES_REQUESTED** [review]

File: src/plexil-adapter/OwInterface.h [hunk 6]
```diff
@@ -97,6 +103,44 @@ class OwInterface
...
+                                        bool b );
+
+//////////////////// FAULTS FOR SYSTEM LEVEL ////////////////////////
+
+  std::map<std::string,std::pair<uint64_t, bool>> systemErrors = 
+  {
+    {"ARM_EXECUTION_ERROR", std::make_pair(4,false)},
+    {"POWER_EXECUTION_ERROR", std::make_pair(512,false)},
+    {"PT_EXECUTION_ERROR", std::make_pair(128,false)}
+  };
```
[the remark this answers is not in the document]
Comment (lanseafood) [reply src/plexil-adapter/OwInterface.h:123]: will do

File: src/plexil-adapter/OwInterface.cpp (hunk 4 above)
Comment (lanseafood) [reply src/plexil-adapter/OwInterface.cpp:329]: it's because that function accepts a uint32 where the system faults expects a uint64. If not for that there could be more optimization I think.

File: src/plexil-adapter/OwInterface.cpp [hunk 7]
```diff
@@ -310,6 +308,97 @@ static void handle_joint_fault (Joint joint, int joint_index,
...
+    bool b = entry.second.second;
+
+    if (checkFaultMessages("ANTENNA", msg_val, key, value, b)) {
+      ptErrors[key].second = !ptErrors[key].second;
+    }
+  }
+}
+
+bool OwInterface::checkFaultMessages
+(string fault_component, uint32_t msg_val, string key, uint32_t value, bool b )
```
Comment (Samahu) [inline src/plexil-adapter/OwInterface.cpp:389]: Use templates to generalize the method for different types of error vectors
```suggestion
template <typename T>
bool OwInterface::checkFaultMessages(
  string fault_component, T msg_val, string key, T value, bool flag )
```
You need to change the method declaration too.

File: src/plexil-adapter/OwInterface.cpp [hunk 8]
```diff
@@ -310,6 +308,97 @@ static void handle_joint_fault (Joint joint, int joint_index,
...
+  if (!b && ((msg_val & value) == value)){
+    ROS_ERROR("%s ERROR: %s", fault_component.c_str(),  key.c_str() );
+    return 1;
+  }
+  else if (b && ((msg_val & value) != value)){
+    ROS_INFO("RESOLVED %s ERROR: %s", fault_component.c_str(), key.c_str() );
+    return 1;
+  }
+  return 0;
+}
```
Comment (Samahu) [inline src/plexil-adapter/OwInterface.cpp:400]: return boolean instead of 0s and 1s

File: src/plexil-adapter/OwInterface.cpp (hunk 4 above)
Comment (Samahu) [reply src/plexil-adapter/OwInterface.cpp:329]: See my notes on refactoring `checkFaultMessages` to use templates

**Review by Samahu — CHANGES_REQUESTED** [review]

**Comment by Samahu** [comment]
I don't know if you still have more plans as part of this feature but at current state when this is merged we will have two ways for receiving faults induced by us. @mikedalal are we going to merge the two mechanisms into one or each mechanism serves a different purpose?

File: src/plexil-adapter/OwInterface.h (hunk 5 above)
Comment (lanseafood) [reply src/plexil-adapter/OwInterface.h]: just tried and subscriber won't compile if we use unique ptr

File: src/plexil-adapter/OwInterface.h [hunk 9]
```diff
@@ -17,6 +17,12 @@
 #include <sensor_msgs/Image.h>
 #include <geometry_msgs/Point.h>
 #include <string>
+#include <utility>
```
[the remark this answers is not in the document]
Comment (lanseafood) [reply src/plexil-adapter/OwInterface.h:20]: will remove!

File: src/plexil-adapter/OwInterface.h (hunk 5 above)
Comment (Samahu) [reply src/plexil-adapter/OwInterface.h]: You may need to read a little on `std::unique_ptr` but in short you should the initialization code to the following:
```bash
m_<...>Subscriber.reset( new ros::Subscriber( ... ) );
```

File: src/plexil-adapter/OwInterface.cpp [hunk 10]
```diff
@@ -569,6 +667,23 @@ void OwInterface::initialize()
     m_guardedMoveSubscriber = new ros::Subscriber
       (m_genericNodeHandle ->
        subscribe("/guarded_move_result", qsize, guarded_move_callback));
+    // subscribers for fault messages
+    m_systemFaultMessagesSubscriber = new ros::Subscriber
+      (m_genericNodeHandle ->
+       subscribe("/system_faults_status", qsize,  
+                &OwInterface::systemFaultMessageCallback, this));
```
Comment (Samahu) [inline src/plexil-adapter/OwInterface.cpp:674]:
```suggestion
    m_systemFaultMessagesSubscriber.reset(new ros::Subscriber
      (m_genericNodeHandle ->
       subscribe("/system_faults_status", qsize,  
                &OwInterface::systemFaultMessageCallback, this)));
```

Comment (Samahu) [reply src/plexil-adapter/OwInterface.cpp:674]: also you need to remove the `delete` operation against raw pointer

File: src/plexil-adapter/OwInterface.cpp [hunk 11]
```diff
@@ -310,6 +308,93 @@ static void handle_joint_fault (Joint joint, int joint_index,
...
+    ROS_ERROR("%s ERROR: %s", fault_component.c_str(),  key.c_str() );
+    return true;
+  }
+  else if (b && ((msg_val & value) != value)){
+    ROS_INFO("RESOLVED %s ERROR: %s", fault_component.c_str(), key.c_str() );
+    return true;
+  }
+  return false;
+}
+
```
Comment (kmdalal) [inline src/plexil-adapter/OwInterface.cpp:394]: It appears that the four previous functions could be replaced by a single template function!

**Comment by kmdalal** [comment]
> I don't know if you have more plans as part of this feature but at current state when this is merged we will have two ways for receiving faults induced by us. @mikedalal are we going to merge the two mechanisms into one or each mechanism serves a different purpose?

For fault _detection_ this branch introduces a new mechanism based on topics, which I think can replace the existing mechanism based on direct inspection of ROS parameters, which was a hack to begin with.

For fault _reporting_, this branch introduces new found/resolved messages which are great.  The existing, continuous messages are still useful, because the new messages can get lost quickly in terminal output.  They are also more detailed.  

To support the existing fault reporting while replacing the existing fault detection, the new fault detection needs to maintain some new state structure for faults. 

All this work could be done in this branch, or as a new Jira...

File: src/plexil-adapter/OwInterface.cpp [hunk 12]
```diff
@@ -517,6 +604,10 @@ OwInterface::~OwInterface ()
   if (m_socSubscriber) delete m_socSubscriber;
   if (m_rulSubscriber) delete m_rulSubscriber;
   if (m_guardedMoveSubscriber) delete m_guardedMoveSubscriber;
+  // if (m_systemFaultMessagesSubscriber) delete m_systemFaultMessagesSubscriber;
+  // if (m_armFaultMessagesSubscriber) delete m_armFaultMessagesSubscriber;
+  // if (m_powerFaultMessagesSubscriber) delete m_powerFaultMessagesSubscriber;
+  // if (m_ptFaultMessagesSubscriber) delete m_ptFaultMessagesSubscriber;
```
Comment (Samahu) [inline src/plexil-adapter/OwInterface.cpp:610]: Why are you still holding on them? :roll_eyes:
```suggestion
```

**Comment by Samahu** [comment]
> > I don't know if you have more plans as part of this feature but at current state when this is merged we will have two ways for receiving faults induced by us. @mikedalal are we going to merge the two mechanisms into one or each mechanism serves a different purpose?
> 
> For fault _detection_ this branch introduces a new mechanism based on topics, which I think can replace the existing mechanism based on direct inspection of ROS parameters, which was a hack to begin with.
> 
> For fault _reporting_, this branch introduces new found/resolved messages which are great. The existing, continuous messages are still useful, because the new messages can get lost quickly in terminal output. They are also more detailed.
> 
> To support the existing fault reporting while replacing the existing fault detection, the new fault detection needs to maintain some new state structure for faults.
> 
> All this work could be done in this branch, or as a new Jira...

Probably a different ticket would be better since @lanseafood may not be as familiar as you with the code to merge things. I'll leave it up to you either way to decide.

**Review by Samahu — APPROVED** [review]

**Review by kmdalal — APPROVED** [review]

File: src/plexil-adapter/OwInterface.cpp (hunk 12 above)
Comment (lanseafood) [reply src/plexil-adapter/OwInterface.cpp:610]: lol forgot to delete!
